src/data/text_input.cpp

```diff
@@ -43,6 +46,9 @@ SentenceTuple TextInput::next() {
     std::string line;
     if(io::getline(dummyStream, line)) {
       Words words = vocabs_[i]->encode(line, /*addEOS =*/ true, /*inference =*/ inference_);
+      if(this->maxLengthCrop_ && words.size() > this->maxLength_) {
+        words.resize(maxLength_);
+      }
       if(words.empty())
         words.push_back(Word::ZERO); // @TODO: What is this for? @BUGBUG: addEOS=true, so this can never happen, right?
```

Comment:
I think the EOS token needs to be added after cropping the sentence, similar to https://github.com/marian-nmt/marian-dev/blob/master/src/data/corpus_base.cpp#L211.
Then that comment with TODO can be removed.